### PR TITLE
Major updates

### PR DESCRIPTION
**Pull request by travisghansen**
Pretty major overhaul. Open to suggestions for sure. Also untested ATM :)

- makes the basic object the `Controller` object
- ensures `station` and `program` objects remain persistent
- `stations` and `programs` are now `dict` instead of `list`
- initiate with actual password instead of md5 password
- initiate with url instead of host (to support https and crazy ports etc)
- support for http basic auth
- can disable ssl verification
- remove `update` call from object instantiation
- rename `update` to `refresh` as update sort of implies making changes

### CODE REVIEW

File: pyopensprinkler/__init__.py [hunk 1]
```diff
@@ -1,84 +1,165 @@
...
+        self._http_client = client
 
     def request(self, path, params=None):
         if params is None:
             params = {}
-        """Make a request from the API."""
+        """Make a request to the API."""
         params["pw"] = self._md5password
         qs = urllib.parse.urlencode(params)
+        url = '{}{}?{}'.format(self._baseUrl, path, qs)
```
Comment (vinteo) [inline pyopensprinkler/__init__.py:47]: seems like the `f"{}"` style is preferred

**Comment by vinteo** [comment]
All great improvements, I agree with `update` -> `refresh`. Shows that a second set of eyes and hands can really improve the codebase :)

File: pyopensprinkler/__init__.py [hunk 2]
```diff
@@ -1,84 +1,165 @@
...
+        """OpenSprinkler Controller initializer."""
 
-    def __init__(self, host, md5password):
-        """OpenSprinkler class initializer."""
-        self._host = host
-        self._md5password = md5password
-        self._baseUrl = f"http://{self._host}"
-        self._programs = []
-        self._stations = []
+        if client_opts is None:
```
Comment (vinteo) [inline pyopensprinkler/__init__.py:20]: Just wondering what the difference is with this and having `clients_opts={}` in the constructor signature?

File: pyopensprinkler/__init__.py [hunk 3]
```diff
@@ -1,84 +1,165 @@
...
 import httplib2
 from backoff import expo, on_exception
 
-from pyopensprinkler.device import Device
 from pyopensprinkler.program import Program
 from pyopensprinkler.station import Station
 
-_HTTP = httplib2.Http()
 
+class Controller(object):
```
Comment (vinteo) [inline pyopensprinkler/__init__.py:14]: `Controller` seems a bit too generic as the main class name, maybe keep `OpenSprinkler` but leave the controller references within the class itself?

File: pyopensprinkler/__init__.py (hunk 2 above)
Comment (travisghansen) [reply pyopensprinkler/__init__.py:20]: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Crazy python stuff.

File: pyopensprinkler/__init__.py (hunk 3 above)
Comment (travisghansen) [reply pyopensprinkler/__init__.py:14]: The library itself is the namespacing right? If someone wants to be more specific they can `import as` if I understand correctly.

File: pyopensprinkler/__init__.py (hunk 1 above)
Comment (travisghansen) [reply pyopensprinkler/__init__.py:47]: I'll update this. Another massive update coming shortly..

File: pyopensprinkler/__init__.py (hunk 3 above)
Comment (vinteo) [reply pyopensprinkler/__init__.py:14]: Ok

File: pyopensprinkler/__init__.py (hunk 2 above)
Comment (vinteo) [reply pyopensprinkler/__init__.py]: Well, I learned something today

File: pyopensprinkler/station.py [hunk 4]
```diff
@@ -1,42 +1,76 @@
...
+        return self._manual_run(params)
+
+    def stop(self):
+        """Stop station"""
+        params = {"en": 0}
+        return self._manual_run(params)
+
+    def toggle(self):
+        """Toggle station"""
+        if self.is_running:
```
Comment (vinteo) [inline pyopensprinkler/station.py:55]: I think there might be an issue here if the station is triggered from elsewhere or the run is finished as `is_running` maybe not have the right state. May have to do a `refresh()` here

File: pyopensprinkler/station.py [hunk 5]
```diff
@@ -48,13 +82,63 @@ def seconds_remaining(self):
...
+    def start_time(self):
+        """Retrieve start time"""
+        return self._get_status_variable(2)
+
+    @property
+    def max_name_length(self):
+        return self._controller._state["stations"]["maxlen"]
+
+    @property
+    def masop(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:95]: with these properties should we have a more user readable names?

Comment (travisghansen) [reply pyopensprinkler/station.py:95]: I'm game. Finishing up a similar set for programs right now..

File: pyopensprinkler/station.py (hunk 4 above)
Comment (travisghansen) [reply pyopensprinkler/station.py:55]: You mean do a `refresh()` before invoking just to ensure non-stale data?

**Comment by vinteo** [comment]
While you are testing can you update the httplib2 library as well?

File: pyopensprinkler/station.py (hunk 4 above)
Comment (vinteo) [reply pyopensprinkler/station.py:55]: yup

Comment (travisghansen) [reply pyopensprinkler/station.py:55]: OK, I'll add it.

**Comment by travisghansen** [comment]
Yeah, planned on that actually.

File: pyopensprinkler/station.py (hunk 4 above)
Comment (travisghansen) [reply pyopensprinkler/station.py:56]: Comitted.

File: pyopensprinkler/station.py (hunk 5 above)
Comment (travisghansen) [reply pyopensprinkler/station.py:96]: Want to review the applicable properties for both stations and programs and propose names?

**Comment by travisghansen** [comment]
OK, this is about what I would expect it to be for an initial stab. We can change names etc but I think it pretty well has the core feature set.

File: pyopensprinkler/station.py [hunk 6]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+        return self._bit_check("masop")
+
+    @property
+    def masop2(self):
+        return self._bit_check("masop2")
+
+    @property
+    def ignore_rain(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:104]: `ignore_rain_delay`

File: pyopensprinkler/station.py [hunk 7]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+        return self._bit_check("ignore_rain")
+
+    @property
+    def ignore_sn1(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:108]: `ignore_rain_sensor_1`

File: pyopensprinkler/station.py [hunk 8]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+        return self._bit_check("ignore_sn1")
+
+    @property
+    def ignore_sn2(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:112]: `ignore_rain_sensor_2`

File: pyopensprinkler/station.py [hunk 9]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+        return self._bit_check("ignore_sn2")
+
+    @property
+    def stn_dis(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:116]: `disabled`

File: pyopensprinkler/station.py [hunk 10]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+        return self._bit_check("stn_dis")
+
+    @property
+    def stn_seq(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:120]: `sequential_operation`

File: pyopensprinkler/station.py [hunk 11]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+        return self._bit_check("stn_seq")
+
+    @property
+    def stn_spe(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:124]: I couldn't figure out what this refers to lol ... maybe `special` 🤷

File: pyopensprinkler/station.py [hunk 12]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+    @property
+    def max_name_length(self):
+        return self._controller._state["stations"]["maxlen"]
+
+    @property
+    def masop(self):
+        return self._bit_check("masop")
+
+    @property
+    def masop2(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:100]: `master_2_operation`

File: pyopensprinkler/station.py [hunk 13]
```diff
@@ -48,13 +83,63 @@ def seconds_remaining(self):
...
+    def start_time(self):
+        """Retrieve start time"""
+        return self._get_status_variable(2)
+
+    @property
+    def max_name_length(self):
+        return self._controller._state["stations"]["maxlen"]
+
+    @property
+    def masop(self):
```
Comment (vinteo) [inline pyopensprinkler/station.py:96]: `master_1_operation`

File: setup.py [hunk 14]
```diff
@@ -18,7 +18,7 @@
     author="Vincent Teo",
```
Comment (vinteo) [inline setup.py:18]: Do you want to add your name here too (comma separated)

**Comment by vinteo** [comment]
I ran the `black` formatter on your branch, looks like the only issue is double quotes for everything instead of single quotes

File: pyopensprinkler/station.py (hunk 6 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: Done.

File: pyopensprinkler/station.py (hunk 7 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: I did `sensor_1_ignored` as the sensor may or may not be a rain sensor.

File: pyopensprinkler/station.py (hunk 8 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: I did `sensor_2_ignored` as the sensor may or may not be a rain sensor.

File: pyopensprinkler/station.py (hunk 9 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: Done.

File: pyopensprinkler/station.py (hunk 10 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: Done.

File: pyopensprinkler/station.py (hunk 11 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: It is special I believe. If the bit is set you can get extra info from the `/je` endpoint from what I can tell in the doc. ie: GPIO or HTTP type stations etc. Renamed to `special`.

File: pyopensprinkler/station.py (hunk 12 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: Done.

File: pyopensprinkler/station.py (hunk 13 above)
Comment (travisghansen) [reply pyopensprinkler/station.py]: Done.

File: setup.py (hunk 14 above)
Comment (travisghansen) [reply setup.py]: Done.

**Comment by travisghansen** [comment]
I changed the formatter to `black` in vscode so hopefully that's better now..

**Comment by travisghansen** [comment]
Still need to update version numbers, docs, etc at this point.

**Comment by vinteo** [comment]
Great! Just need to update docs and test. Thanks so much again

**Comment by travisghansen** [comment]
Yup. Hopefully you can run with it on hass side...particularly on the GUI bits. I can probably help a bit with the plumbing though in hass and definitely code review and testing.

If you do should we move the hass plugin to it’s own repo (similar to the hass-opensprinkler repo or even in that repo in a new branch or something) until it’s merged? Might be easier for testing/collaborating.

**Comment by vinteo** [comment]
Yes I have been thinking for a now now bring the custom component (hass-opensprinkler) up to speed with the PR I have for hass since it will probably take a while for the official integration to get in and like you said its easier for wider testing etc. But just haven't gotten around to it yet.

**Comment by travisghansen** [comment]
Yeah ok. I guess just shout out when you've got it moved over or something. I forked it originally and started hacking it like crazy in my own repo but probably easiest at this point to trash that and work jointly on it in a dev branch in your main repo for now.

I'm pretty motivated to get it over the finish line at this point as it's one of the few remaining items for my hass setup to get buttoned up.

**Comment by travisghansen** [comment]
Poking at docs a little bit now. Hopefully I'll have those fleshed out sometime tomorrow and we'll do a final review before snapping a release.

File: pyopensprinkler/__init__.py [hunk 15]
```diff
@@ -1,84 +1,213 @@
...
+        """Retrieve rain delay stop time"""
+        return self._get_variable("rdst")
+
+    @property
+    def rain_sensor_enabled(self):
+        """Retrieve rain sensor enabled"""
+        try:
+            return bool(self._get_variable("rs"))
+        except KeyError:
+            return False
```
Comment (vinteo) [inline pyopensprinkler/__init__.py:189]: I think maybe this should return `None` instead

**Comment by vinteo** [comment]
I just did some testing, all the properties seem to work but I couldn't get the auto refresh to work.

```python
os = Controller(...) # also tried os = OpenSprinkler(..., {"auto_refresh_on_update": True})
os.refresh()
os.station[0].run() # works, station runs
os.station[0].status # still shows 'idle' even though station is running
os.refresh()
os.station[0].status # now shows correct value 'once_program'
```

Can you check?

**Comment by travisghansen** [comment]
Yeah I’ll have a look.

**Comment by travisghansen** [comment]
Should we use something like this for the docs? https://pypi.org/project/pydoc-markdown/

File: pyopensprinkler/__init__.py (hunk 15 above)
Comment (travisghansen) [reply pyopensprinkler/__init__.py:189]: This is for the legacy stuff yeah? I’d be OK with that..

**Comment by travisghansen** [comment]
Ok, station issue is what I feared, literally os it too slow and doesn't return the proper values yet because it's happening so quickly. We can put a `sleep(1)` or something in there I suppose although I'm not entirely sure that's the right idea either.

File: pyopensprinkler/__init__.py (hunk 15 above)
Comment (travisghansen) [reply pyopensprinkler/__init__.py]: Done.

**Comment by travisghansen** [comment]
```python
        update_paths = ["/cv", "/co", "/cs", "/cm", "/mp", "/cp", "/dp", "/up", "/cr"]
        if self._opts["auto_refresh_on_update"]["enabled"] and path in update_paths:
            #  .1 was not enough settle time
            # .25 was mostly good but still too fast at times
            #  .5 was mostly good but still too fast at times
            # .75 was mostly good but still too fast at times
            #   1 was consistently enough time
            time.sleep(float(self._opts["auto_refresh_on_update"]["settle_time"]))
            self.refresh()
```

**Comment by vinteo** [comment]
I'm happy to go with pydoc-markdown

**Comment by travisghansen** [comment]
Messing with it right now and can't seem to get the output sane (or at least what I would expect to be sane).

**Comment by travisghansen** [comment]
I give up with the auto-generation tools. If you want to try it out that'd be cool, otherwise I guess we'll just hand-write docs or provide a sample file. Something like this:

```python
from pyopensprinkler import Controller as OpenSprinklerController

# instantiate controller
controller = OpenSprinklerController("http://host", "password")
# initial load of state
controller.refresh()

# controller
if True:
    # properties
    print(f"##### starting controller info dump")
    print(f"firemware version: {controller.firmware_version}")
    print(f"hardware version: {controller.hardware_version}")
    print(f"last run station: {controller.last_run_station}")
    print(f"last run program: {controller.last_run_program}")
    print(f"last run duration: {controller.last_run_duration}")
    print(f"last run end time: {controller.last_run_end_time}")
    print(f"rain delay enabled: {controller.rain_delay_enabled}")
    print(f"rain delay stop time: {controller.rain_delay_stop_time}")
    print(f"rain sensor enabled: {controller.rain_sensor_enabled}")
    print(f"sensor 1 enabled: {controller.sensor_1_enabled}")
    print(f"sensor 2 enabeld: {controller.sensor_2_enabled}")
    print(f"operation enabled: {controller.operation_enabled}")
    print(f"water level: {controller.water_level}")
    print(f"stations count: {len(controller.stations)}")
    print(f"programs count: {len(controller.programs)}")

    # methods
    if False:
        controller.disable()
        controller.enable()

# stations
if False:
    # properties
    for i, station in controller.stations.items():
        print(f"##### starting station {i} info dump")
        print(f"name: {station.name}")
        print(f"index: {station.index}")
        print(f"is runnng: {station.is_running}")
        print(f"running program id: {station.running_program_id}")  # 1-indexed
        print(f"seconds remaining: {station.seconds_remaining}")
        print(f"start time: {station.start_time}")
        print(f"max name length: {station.max_name_length}")
        print(f"master 1 operation enabled: {station.master_1_operation_enabled}")
        print(f"master 2 operation enabled: {station.master_2_operation_enabled}")
        print(f"rain delay ignored: {station.rain_delay_ignored}")
        print(f"sensor 1 ignored: {station.sensor_1_ignored}")
        print(f"sensor 2 ignored: {station.sensor_2_ignored}")
        print(f"disabled: {station.disabled}")
        print(f"sequential operation: {station.sequential_operation}")
        print(f"station type: {station.station_type}")
        print(f"status: {station.status}")

        # methods
        if False:
            station.run()
            station.stop()
            station.toggle()

# programs
if False:
    # properties
    for i, program in controller.programs.items():
        print(f"##### starting program {i} info dump")
        print(f"name: {program.name}")
        print(f"index: {program.index}")
        print(f"enabled: {program.enabled}")
        print(f"is running: {program.is_running}")
        print(f"use weather adjustments: {program.use_weather_adjustments}")
        print(f"odd even restriction: {program.odd_even_restriction}")
        print(f"odd even restriction name: {program.odd_even_restriction_name}")
        print(f"program schedule type: {program.program_schedule_type}")
        print(f"program schedule type name: {program.program_schedule_type_name}")
        print(f"start time type: {program.start_time_type}")
        print(f"start time type name: {program.start_time_type_name}")

        # methods
        if False:
            program.run()
            program.enable()
            program.disable()
```

**Comment by vinteo** [comment]
The library is pretty simple so I think we can just update the README and add your example file and we are good to go

**Comment by travisghansen** [comment]
Yeah agreed. Should we draw up a clear idea (here or on the other project) about how we want all this represented in hass? ie: do we want to expand the available methods to set the various enabled/ignored properties so they can be manifest as switches instead of binary sensors etc etc

**Comment by vinteo** [comment]
I think we should start with the basics which we already have, If there are requests to add switches for those (or even the sensors) we can add them as we go. I don't think we should flood HASS with a ton of sensors and switches. When we add them we can include configs to filter unwanted ones

**Comment by travisghansen** [comment]
Yeah I especially don’t want to double up like we have now (binary sensor and switch entities for the same thing). Maybe let’s map out exactly what we want it to look like and just include those data points as device attrs in hass so people can see them but not interact with them beyond the basic switches?

**Comment by vinteo** [comment]
This is what I was thinking:
- sensors: station status with other station properties as attributes
- switches: controller operation (with controller properties as attributes, water level, last run etc), program operation (with program properties as attributes)

No more binary sensors, scenes or switches for stations, instead:

- services: run program, run station (can provide run seconds as data)

I am considering water level might be worth it as its own sensor though, but I don't use it not sure if any other people have use cases for it. But in the end they can still use a sensor template.

**Comment by travisghansen** [comment]
I like the direction a lot. Should we do switches for stations as well (to be able to turn off a running station)? One trouble with that is what do we do for turn on in terms of duration?

Hypothetically water level could be a light allowing the user to alter the current level. I’m not sure what the upper and lower bounds are for it (and not sure what on/off should do).

**Comment by vinteo** [comment]
If the user wants a switch for station, they can create it using a switch template using the run/stop service and its up to them how they want to provide the run time, through an input number for example. that way the user isn't locked down but have the flexibility to still do it. 

I am not a fan of using platforms for stuff they were menat to bem for water level, we can also have a service to set water levels and a sensor for it then the user can create their own entities

I will create some issues in the hass-opensprinkler repo so we can keep track of these, feel free to assign yourself to the ones you want to work on

**Comment by travisghansen** [comment]
Yeah ok cool. I’m pretty new to hass so not entirely sure the best way to do things yet but mixing with provided services seems like a great route! We could probably add a few more services over time to allow even greater management capabilities.

**Comment by vinteo** [comment]
Should we merge this and do a release since the release doesn't need the docs? The README update can be a new PR.

**Comment by travisghansen** [comment]
I’ll update it shortly, then we can merge.

**Comment by travisghansen** [comment]
OK, implemented pretty much everything else at this point. I'm good to merge when you are. Probably a final review would be good since I've made some minor adjustments to property names etc.

**Comment by vinteo** [comment]
Great work on the new methods. Merging...

**Comment by travisghansen** [comment]
Huzzah! Now to the fun part :)

**Pull request merged**